### DISCUSSION TITLE
Proposal for an Additional Non-httpOnly Cookie in CookieTransport

### DISCUSSION THREAD

**Discussion by gyolkin** [fastapi-users/fastapi-users#1324] [Ideas]
The [CookieTransport class](https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/authentication/transport/cookie.py) allows setting an httpOnly cookie for authentication purposes. However, it becomes impossible for the frontend to check the existence of the authentication cookie when the httpOnly flag is set (as we cannot access this cookie using JS). This becomes a problem when the client needs to determine if the user is already authenticated.

### Proposal
I propose to solve this problem by creating an additional non-httpOnly cookie that could be set alongside the main authentication cookie. This cookie would serve as an indicator for the frontend to recognize the user's authentication state (e.g., at the application initialization stage).

### Implementation Suggestion
A new optional parameter could be added to the CookieTransport. When enabled, we would set an additional non-httpOnly cookie with similar lifespan parameters as the main authentication cookie. Remove it when logging out.

I am ready to contribute by creating a PR with the necessary changes if the idea is accepted.

**Comment by frankie567** [comment]
Interesting, thank you for your proposal! I'm curious to see a PR implementing this pattern :)

**Comment by 10-mahmoud** [comment]
Hah! Came here looking for exactly this. Agreed with all of the above. Sharing some of my own notes below:

I haven't dug deep into the code, but in trying to implement this with the existing API, it struck me as odd that `FastAPIUsers` accepts a sequence of backends, but, e.g., `get_oauth_router()` can only accept one backend. So even if you have two backends, one with a secure cookie and another for UI/convenience, the router won't set both.

In my use case, which I'm approaching as a custom middleware, the secure cookie is a signed/dated UUID (server session key, DB-backed, `samesite=strict`), while the JS-accessible cookie is JWT/PASETO (basic auth state, not persisted, `samesite=lax`).

**Reply by frankie567** [reply]
If you have several authentication backends, you need to have a router for each one of them. It makes sense since that depending on the context, you'll need to authenticate using one method or the other.

**Reply by 10-mahmoud** [reply]
Yeah, I understand that's how the library is factored now. Just found it odd that `FastAPIUsers` wants to know about the backends, and then the `get_router()` call passed the backend again. I suppose we all could use a reminder from time to time, and `get_router` is no different.

**Reply by frankie567** [reply]
The reason is that there are two "sides" for an authentication backend:

1. Check if a request is authenticated (i.e. look for a valid token/cookie/whatever). That's why the user manager needs to know *every* backends available, so it can test all of them when a request comes.
2. Generate a fresh token/cookie/whatever given valid user credentials. That's the role of the auth routers, and since each backend can have its own way of returning the fresh token/cookie/whatever, a single router is needed per backend.

The happy consequence of this architecture is that you could very well have a configuration where the manager checks if a token is valid, but without a router to generate it (maybe it's generated by another microservice). Or the contrary!

**Reply by 10-mahmoud** [reply]
I appreciate you sharing those details! I might have skimmed past that section in the docs. :sweat_smile: If that clarification isn't in there, I think it'd make a worth addition/FAQ. Thanks!
